### DISCUSSION TITLE
milvus insertion speed problem

### DISCUSSION THREAD

**Discussion by ingale726** [milvus-io/milvus#21338] [Q&A and General discussion]
I tried to insert data into milvus concurrently and found that the speed was 30 bars/second. Is this a normal value? Is there any code to speed up insertion?

**Comment by ingale726** [comment]
```
def connect_milvus():
    from pymilvus import (
    connections,
    utility,
    FieldSchema,
    CollectionSchema,
    DataType,
    Collection,
    )
    connections.connect("default", host="localhost", port="19530")
    hello_milvus = Collection("embeddings")
    return hello_milvus
```

I use the variable hello_milvus for concurrent insertion. Am I doing something wrong? Should I create 10 variables like this and use each variable for concurrent insertion?

**Reply by yhmo** [reply]
No need to add more concurrents to insert, the bottleneck is not client side. All the insert data will be processed by proxy node, split into parts according to shard number, and send these parts into Pulsar, datanode and querynode subscribe the data from Pulsar.
Typically, for 128-dim vectors, insert batch by batch, each batch contains 10000 vectors, the average speed is about 30000 ~ 50000 vectors per second.

**Reply by ingale726** [reply]
How to implement batch insertion?
